include/revng/StackAnalysis/ABI/CallingConventionTrait.h

```diff
@@ -0,0 +1,645 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/Model/Binary.h"
+#include "revng/Model/Register.h"
+#include "revng/Model/Type.h"
+
+namespace abi {
+
+/// A trait struct which is specialized for all of the supported ABIs.
+/// It describes the specifics of the calling convention for the ABI.
+template<model::abi::Values V>
+struct CallingConventionTrait {
+
+  /// States whether specified ABI allows using both generic and vector
+  /// registers in parallel.
+  /// For an example of an ABI with position based registers see Microsoft x64
+  /// documentation, for one without - see SystemV x64 documentation.
+  static constexpr bool ArgumentsArePositionBased = false;
+
+  /// States whether aggregate objects (structs, unions, etc) are allowed to
+  /// be passed around in registers.
+  ///
+  /// \note: When `AllowPassingAggregatesInRegisters` is false, but either
+  /// `GenericRegistersUsedPerAggregateArgument` or
+  /// `GenericRegistersUsedPerAggregateReturnValue` is non-zero, their values
+  /// are used for determining the maximum number of registers to use for
+  /// primitive values that are bigger than a generic registers but don't
+  /// use vector registers (for example `__int128`).
+  static constexpr bool AllowPassingAggregatesInRegisters = false;
+
+  /// States whether specified ABI allows accepting aggregate object arguments
+  /// (structs, unions, etc) in the designated registers if they provide enough
+  /// space (including padding) and the maximum number of registers allowed per
+  /// aggregate argument.
+  ///
+  /// `0` means that using registers for aggregate objects is disallowed.
+  /// `1` means that passing aggregate values are only allowed if their size
+  /// (including padding) doesn't exceed the size of a single generic register.
+  /// `2` means that passing aggregate values are only allowed if their size
+  /// (including padding) doesn't exceed the size of two generic registers.
+  /// ...
+  /// `-1` means that there are no limitations and all the allowed registers
+  /// could be used for a single aggregate argument.
```

Comment:
`-1` for unsigned types is a bit weird. Make a named `constexpr` value using `std::numeric_limits<size_t>::max()`.